packages/qwik-city/runtime/src/server-functions.ts

```diff
@@ -230,6 +231,58 @@ export const validatorQrl = ((
  */
 export const validator$: ValidatorConstructor = /*#__PURE__*/ implicit$FirstArg(validatorQrl);
 
+type ValibotObjectSchema = v.ObjectSchema<v.ObjectShape>;
+type ValibotObjectShapeOrSchema = v.ObjectShape | ValibotObjectSchema;
+
+/**
+ * @public
+ */
+export const valibotQrl = (
+  qrl: QRL<ValibotObjectShapeOrSchema | ((z: any, ev: RequestEvent) => ValibotObjectShapeOrSchema)>
+): DataValidator => {
+  if (isServer) {
+    return {
+      async validate(ev, inputData) {
+        const schema: Promise<ValibotObjectSchema> = qrl.resolve().then((obj) => {
+          if (typeof obj === 'function') {
+            obj = obj(z, ev);
+          }
+          if (obj.schema === 'object') {
+            return obj as ValibotObjectSchema;
+          } else {
+            return v.object(obj as v.ObjectShape);
+          }
+        });
+
+        const data = inputData ?? (await ev.parseBody());
+        const result = v.safeParse(await schema, data);
+        if (result.success) {
+          return result;
+        } else {
+          if (isDev) {
+            console.error(
+              '\nVALIDATION ERROR\naction$() Valibot validation failed.',
+              '\n  - Issues:',
+              result.issues
+            );
+          }
+          return {
+            success: false,
+            status: 400,
+            error: v.flatten(result.issues),
```

Comment:
Should I return a `fieldErrors` key to match the existing Zod implementation?

Reply:
I’d rather use Valibot’s return value as is, but I wanted to run it by someone just to check. Thanks.